casper/src/main/resources/RevVault.rho

```diff
@@ -0,0 +1,154 @@
+                    match eitherVault {
+                      ("Right", vault) => vaultMapStore!(vaultMap.set(key, vault))
+                      ("Left",  _    ) => vaultMapStore!(vaultMap)
+                    } |
+                    ret!(eitherVault)
+                  }
+                }
+              }
+              value => {
+                vaultMapStore!(vaultMap) | ret!(("Right", value))
+              }
+            }
+          }
+        } |
+
+        contract _makeVault(@ownerPubKeyBytes, @initialAmount, ret) = {
+
+          new revAddrCh, eitherRevAddrCh, purseCh, eitherPurseCh in {
+            RevAddress!("fromPublicKey", ownerPubKeyBytes, *revAddrCh) |
+            @Either!("fromNillable <-", *revAddrCh, "Invalid public key", *eitherRevAddrCh) |
+            mint!("makePurse", initialAmount, *purseCh) |
+            @Either!("fromNillable <-", *purseCh, "Couldn't create purse", *eitherPurseCh) |
+            @Either!("map2 <-", *eitherRevAddrCh, *eitherPurseCh, *_newVault, *ret)
+          }
+
+        } |
+
+        contract _newVault(@ownRevAddress, purse, ret) = {
+          new revVault in {
+
+            ret!(bundle+{*revVault}) |
+
+            contract revVault(@"balance", ret) = {
+              purse!("getBalance", *ret)
+            } |
+
+            contract revVault(@"transfer", @revAddress, @amount, ret) = {
+
+              new
+                revAddressValid, revAddressValidEither, amountNonNegative, split,
+                eitherPurseCh, purseReadyCh, doDeposit
+              in {
+                RevAddress!("validate", revAddress, *revAddressValid) |
+                @Either!("fromNillableError <-", *revAddressValid, *revAddressValidEither) |
+                @Either!("fromBoolean", amount >= 0, "Amount must be non-negative", *amountNonNegative) |
+                @Either!("flatMap <-", *amountNonNegative, *split, *eitherPurseCh) |
+                for (_, ret <- split) {
+                  new amountPurseCh in {
+                    purse!("split", amount, *amountPurseCh) |
+                    @Either!("fromSingletonList <-", *amountPurseCh, "Insufficient funds", *ret)
+                  }
+                } |
+                @Either!("productR <-", *revAddressValidEither, *eitherPurseCh, *purseReadyCh) |
+                @Either!("flatMap <-", *purseReadyCh, *doDeposit, *ret) |
+                for (@p, retCh <- doDeposit) {
+                  @{revAddress | bundle0{*_revVault}}!("_deposit", p, *retCh)
+                }
+              }
+
+            } |
+
+            contract @{ownRevAddress | bundle0{*_revVault}}(@"_deposit", depositPurse, retCh) = {
+
+              new amountCh, depositSuccessCh in {
+                depositPurse!("getBalance", *amountCh) |
+                for (@amount <- amountCh) {
+
+                  purse!("deposit", amount, *depositPurse, *depositSuccessCh) |
+                  @Either!("fromBoolean <-", *depositSuccessCh, "BUG FOUND: purse deposit failed", *retCh)
```

Comment:
The purse deposit can fail in case of overflow. In such situation we need to revert the split.

Reply:
Yeah, we should at least change the message. I'm not sure though what should happen when one transfers to an already-full ~~wallet~~ vault:
 - should the funds be lost (easiest one)
 - should they be refunded (can cause a cascade and is subject to concurrency bugs)
 - should the vault keep a pile of purses and prevent overflow proactively? (require substantial development effort)

Given that it's unclear how to proceed, and that this is admittedly a rare scenario - are you OK with me scheduling a JIRA for fixing that later?

Reply:
Indeed, this is not trivial fix so I'm ok with a new JIRA issue